### PR TITLE
docs: [FC-0074] add comparison table between the framework and forking openedx

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
### Description
This PR adds a comparison table between extending the platform using events/filters and modifying it directly in a fork.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @mariajgrimaldi!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/docs-openedx-org-maintainers`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: source/developers/concepts/hooks_extension_framework.rst
Comment (sarina) [inline source/developers/concepts/hooks_extension_framework.rst:157]:
```suggestion
| **Community Compatibility**                                        | High, as the community encourages reusable implementations and the contribution of new extension points.       | Low, as changes in forks are usually highly customized and not designed for others to reuse.                                           |
```

File: source/developers/concepts/hooks_extension_framework.rst
Comment (sarina) [inline source/developers/concepts/hooks_extension_framework.rst:159]:
```suggestion
| **Implementation Effort**                                          | Low to medium, depending on the complexity of the receiver logic and filter pipelines.                        | High, as modifications require expertise in the platform and increases technical debt.                                              |
```

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst:142]: I just realized I missed the word plugins in _by leveraging plugins and configurations_

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst:152]: When implementing new features in plugins, even though you could be using an extension point which should provide enough data about what's being extended, e.g., the enrollment filter provides the `user`, `course_key`, and `enrollment mode`, we might want to perform operations on the data like getting the `course overview` object that inevitably could result in importing edx-platform code into plugins. This is an antipattern, and it's strongly not recommended, but there's no current alternative for doing that kind of operation (this might be solved by learning core in the long term). Is this too specific to add here?

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst:154]: I think both the framework and forking could behave similarly under this category cause this mainly depends on how performant the solution is. I will remove this category and move the _it may require additional testing..._ from the fork column to the implementation effort row.

File: source/developers/concepts/hooks_extension_framework.rst
Comment (sarina) [reply source/developers/concepts/hooks_extension_framework.rst:179]: How does this sound as a summary / a little bit more detail & clarity?

```suggestion
|                                                                    | - Although it is not recommended to do so, a plugin might introduce an edx-platform dependency, mainly when using filters, when there is no alternative for getting certain types of data.                         |                                                                                                                                    |
```

Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst]: Yes! Thank you. I implemented the suggestion here: https://github.com/openedx/docs.openedx.org/pull/617/commits/d4266da384e73d2dd6713fe1a897503a19c0ab4c

**Review by Ian2012 — APPROVED** [review]

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:162]: This approach is in contrast to forking -> sounds a bit weird to me.

How about: This approach stands in opposition to the practice of forking ... ?

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:168]:
```suggestion
|                                                                    | core codebase by leveraging plugins and configurations.                                                   |                                                                                                                                    |
```

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:172]:
```suggestion
|                                                                    | - Enables reusable and configurable behavior across the platform.                                             | - Suitable for extremely customized use cases.                                                                                        |
```
I think that hooks framework already allows for high customization. IMHO the modification must be extreme to warrant forking where hooks can do it. The more regular reason is when the platform does not support hooks in the place you want to customize.

Comment (felipemontoya) [reply source/developers/concepts/hooks_extension_framework.rst:172]: Another point in this line would be:

> Enables reusable and configurable behavior across the platform

I think is worth splitting in two:

- Enables reusable behavior across the ecosystem and community.
- Enables configurable behavior across the platform

**Review by sarina — APPROVED** [review]

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:175]: I think we should call it "separation from the upstream project" because core has a different meaning here.

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:178]: > Requires deep understanding of the platform to modify it correctly

I would argue that modifying correctly means using the open-closed principle which is not being used if the road was a fork. 

I would list however the "not being able to reuse by the community" as part of the disadvantages.

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [inline source/developers/concepts/hooks_extension_framework.rst:182]: > Low, as changes in forks are usually highly customized not designed for others to reuse.

And discoverability is low. Most probably no one will ever know "you did that".

File: source/developers/concepts/hooks_extension_framework.rst
Comment (felipemontoya) [reply source/developers/concepts/hooks_extension_framework.rst:178]: Aso a disadvantage of fork:

- difficult to test in isolation

**Review by felipemontoya — COMMENTED** [review]
Having a table of pros and cons is difficult as it is very opinionated. I do think the information within is very accurate. And I commend the work put into this.

Left some comments that reflect my opinion in specific lines. None of them are blocking.

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst:168]: Thanks. I completely missed that.

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst]: > I think that hooks framework already allows for high customization. IMHO the modification must be extreme to warrant forking where hooks can do it. The more regular reason is when the platform does not support hooks in the place you want to customize.

I agree. I added a note suggesting to look into how the framework can help before forking: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R193

I also split the sentence you mentioned into two for more clarity: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R172-R173

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst]: That's more clear. Thank you.

https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R177

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst]: > I would argue that modifying correctly means using the open-closed principle which is not being used if the road was a fork.

I removed the "correctly" from the sentence. I do think that to modify the platform, you need a deep understanding of the components you'd be modifying, mainly because of the implications of the changes. But one can argue that you'd need the same to add new extension points, so I'll drop the line altogether. 

> I would list however the "not being able to reuse by the community" as part of the disadvantages.

Added! Thanks https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R181

File: source/developers/concepts/hooks_extension_framework.rst
Comment (mariajgrimaldi) [reply source/developers/concepts/hooks_extension_framework.rst]: I added a note about discoverability here: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R185

**Comment by mariajgrimaldi** [comment]
@felipemontoya @sarina: thank you for the review!

Felipe, I think I addressed all of your comments with these two commits: 
https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd
https://github.com/openedx/docs.openedx.org/pull/617/commits/125160543a9dc71125d4e9f81135012a487ee82e

**Review by felipemontoya — APPROVED** [review]
Thanks for addressing all the comments. Lgtm

**Comment by mariajgrimaldi** [comment]
Thank you all!

@sarina: merge when you're ready :)

**Pull request merged**
